### DISCUSSION TITLE
Protocol WebSocket

### DISCUSSION THREAD

**Discussion by BorisSutin** [rerdavies/pipedal#188] [General]
Hello Robin. 
      I'm working on the websocket protocol and I have a question. When changing the pedalboard on the client, I get the message "onPedalboardChanged", this is expected, but after it comes the message onPresetsChanged in which I get a list of all pedalboards. And this happens every time the pedalboard is changed. Maybe this method is left over from older versions. Can you explain what onPresetsChanged is used for. When I add a new pedalboard, then this message is expected. But after changing the pedalboard, the receipt is not entirely clear to me.

**Comment by rerdavies** [comment]
Not at my computer right now. But there are two pieces that need to be
replicate to the client model: the value of the current/default selected
present; and whether the current present has been modified (whether a *
needs to be displayed after the preset name). I don't think the state of
the modified flag is carried in that data structure, but the id of the
current selected preset definitely is.

If you haven't found out already, you will discover that tunneling new
messages from client to server and back again is is fairly laborious. It's
probably a case of using an existing message rather than creating a new one
to transmit just the instanceid of the selected preset. it is admittedly a
largish message to send when a uint64_t would have sufficed. But it's not
as if preset selection occurs a 100 times a second. So also a case of
"Don't optimize unless you have to".

Something to keep in mind if you're introducing new messages is that, by
design, Pipedal is designed to support multiple clients connected
simultaneously. So any change to server state has to be echoed back to the
attached clients. And the client has to be prepared to accept and process
notification of changes that it just sent. In very early stages of
implementation, all notifications carried the client id of the client that
initiated the change on the theory that clients might want to shortcut
updates for changes they had sent themselves. In practice, this never
happened; and I'm not actually sure that it would be safe to do so. There
are definitely race conditions where two clients update the same piece of
server state currently. To safely replicate server state, clients must
process the server notification even if it is in response to the a state
change. It is safe for the client to update it's local model state before
it sends a state change to the server. However, to safely maintain
synchronized state with the server when multiple clients are active, server
notifications have to be considered authoritative and final. So don't
bother including a client id

There are definitely of cases where things get challenging. E.g. client one
send a preset change; client 2 sends a control change. The server receives
the preset change before it receives the control change: result: the
control change may get applied to the wrong preset. Although I think
truthfully the current server implementation notices that instance ID of
the control change doesn't match the instance ID of the current loaded
preset.

The whole approach is probably not totally theoretically correct. But I
think broadly the approach is sound. The server processes messages in the
order that it receives them, and then sends notifications to all connected
clients to ensure that they can replicate the server state accurately.

 So do be careful.
